### PR TITLE
implement copy snapshot API for aws and azure

### PR DESCRIPTION
**Pull request by NishanthSpShetty**
# Description

- Allow user to copy existing snapshot into new snapshot.
- Add names in aws volumes and snapshot while creating as its currently empty

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# why
- allow copy of snapshot.
- Add names in aws volumes and snapshot while creating as its currently empty

# How Has This Been Tested?
- manually copy snapshot in aws and azure

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pkg/service/labels/labels.go [hunk 1]
```diff
@@ -43,7 +42,7 @@ func GetNodeLabel(nodeSpec *proto.NodeSpec) map[string]*string {
 }
 
 func ScopeTag() string {
-	return fmt.Sprintf("nb-%s", config.Get().Env)
+	return fmt.Sprintf("nb-%s", "dev")
```
Comment (SscSPs) [inline pkg/service/labels/labels.go:45]: I think this was not intended

File: pkg/service/aws/volume.go [hunk 2]
```diff
@@ -317,3 +332,65 @@ func (a *awsController) DeleteSnapshot(ctx context.Context, req *proto.DeleteSna
...
+		return nil, errors.Wrap(err, "faild to fetch source snapshot")
+	}
+
+	if len(snap.Snapshots) == 0 {
+		a.logger.Error(ctx, "failed to fetch the snapshot, got 0 snapshots in response", "snapshotid", snapshotId)
+		return nil, errors.New("CopySnapshot: failed to fetch the source snapshot")
+	}
+	copyDesc := fmt.Sprintf("copy of snapshot: %s", snapshotId)
+
+	name := fmt.Sprintf("copy-%s", snapshotId)
```
Comment (SscSPs) [inline pkg/service/aws/volume.go:365]: Should we have this? we should probably not hardcode it like this, probably a tag would be better, for source snapshot id(and uri?) 
same comment for other providers

File: pkg/service/aws/volume.go [hunk 3]
```diff
@@ -171,7 +182,11 @@ func (svc awsController) CreateSnapshot(ctx context.Context, req *proto.CreateSn
 	}
 
 	tags := awsTags(labels)
-
+	name := fmt.Sprintf("snapof-%s", volumeid)
```
Comment (SscSPs) [inline pkg/service/aws/volume.go:185]: are we using the name format somewhere to identify something in any place? if yes, can this break backwards compatibility?

Comment (NishanthSpShetty) [reply pkg/service/aws/volume.go:185]: no, just for reference and ease of spotting in web console. No use.

File: pkg/service/labels/labels.go (hunk 1 above)
Comment (NishanthSpShetty) [reply pkg/service/labels/labels.go]: removed.

File: proto/netbookai/spawner/spawner.proto [hunk 4]
```diff
@@ -554,3 +555,17 @@ message DeleteRoute53RecordsRequest {
...
 
 message DeleteRoute53RecordsResponse {}
+
+message CopySnapshotRequest {
+  string provider = 1;
+  string region = 2;
+  string accountName = 3;
+  string snapshotId = 4;
+  string snapshotUri = 5;
+  map<string, string> labels = 6;
```
[the remark this answers is not in the document]
Comment (NishanthSpShetty) [reply proto/netbookai/spawner/spawner.proto:565]: good point, might mislead if the data is being used for different purpose, ex: snapshot was created with result of some experiment or by workspace, and we are using it in some place.
Depends on what exactly we will be doing with the copy I guess.

File: pkg/service/azure/snapshot.go [hunk 5]
```diff
@@ -194,3 +194,61 @@ func (a *azureController) deleteSnapshot(ctx context.Context, req *proto.DeleteS
...
+		a.logger.Error(ctx, "failed to get the azure credentials", "error", err)
+		return nil, errors.Wrap(err, "copySnapshot")
+	}
+
+	sc, err := getSnapshotClient(cred)
+	if err != nil {
+		a.logger.Error(ctx, "faied to get the snapshot client", "error", err)
+		return nil, errors.Wrap(err, "copySnapshot")
+	}
+	name := fmt.Sprintf("copy-%s", req.SnapshotId)
```
Comment (sid255) [inline pkg/service/azure/snapshot.go:213]: Can we pick this from a common place for all providers?

Comment (NishanthSpShetty) [reply pkg/service/azure/snapshot.go:213]: there is a follow up todo, we will address it in that PR.

**Review by sid255 — APPROVED** [review]
LGTM

File: proto/netbookai/spawner/spawner.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (sid255) [reply proto/netbookai/spawner/spawner.proto:565]: We can add a flag but I don't see us using it. So we can add it later when the use case arises.

[the remark this answers is not in the document]
Comment (SscSPs) [reply proto/netbookai/spawner/spawner.proto:565]: Yeah, we can add it if needed, may be as mask, but not needed for now

**Review by SscSPs — APPROVED** [review]

**Pull request merged**